compact.go

```diff
@@ -660,49 +668,66 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 		}
 
 		for i, chk := range chks {
-			if chk.MinTime < meta.MinTime || chk.MaxTime > meta.MaxTime {
-				return errors.Errorf("found chunk with minTime: %d maxTime: %d outside of compacted minTime: %d maxTime: %d",
-					chk.MinTime, chk.MaxTime, meta.MinTime, meta.MaxTime)
+			// Current use cases don't require handling chunks before the requested min time.
+			if chk.MinTime < mint {
+				return BlockStats{}, errors.Errorf("found chunk with minTime: %d maxTime: %d outside of compacted minTime: %d maxTime: %d",
+					chk.MinTime, chk.MaxTime, mint, maxt)
 			}
 
-			if len(dranges) > 0 {
-				// Re-encode the chunk to not have deleted values.
-				if !chk.OverlapsClosedInterval(dranges[0].Mint, dranges[len(dranges)-1].Maxt) {
-					continue
-				}
-				newChunk := chunkenc.NewXORChunk()
-				app, err := newChunk.Appender()
-				if err != nil {
-					return err
-				}
+			// During snapshotting an append to a head's chunk can happen,
+			// so samples outside of the requested time range are expected and need to be removed.
+			if chk.MinTime > maxt {
+				// Remove chunk that is completely outside.
+				chks = append(chks[:i], chks[i+1:]...)
+				continue
+			}
 
-				it := &deletedIterator{it: chk.Chunk.Iterator(), intervals: dranges}
-				for it.Next() {
-					ts, v := it.At()
-					app.Append(ts, v)
-				}
+			if chk.MaxTime > maxt {
+				// Rewrite chunk to remove samples that are outside requested time range.
+				dranges = append(dranges, Interval{Mint: maxt + 1, Maxt: chk.MaxTime})
+				sort.Slice(dranges, func(i, j int) bool {
+					return dranges[i].Mint < dranges[j].Maxt
+				})
+			}
+
+			if len(dranges) == 0 || !chk.OverlapsClosedInterval(dranges[0].Mint, dranges[len(dranges)-1].Maxt) {
+				continue
+			}
+
+			// Re-encode the chunk to not have deleted values or to match requested max time.
```

Comment:
```suggestion
			// Rewrite chunk to remove deleted values.
```

Reply:
HM.. not only deleted values. Also those which are out of max time right?

Reply:
that is right, but in regards to this block it removes samples added to the `dranges` so semantically it removes data from the dranges or in other words remove deletes values(dranges) from the chunk

Reply:
maybe `// Rewrite chunk to remove drange values.`

Reply:
:+1: